common/quotas/clocked_rate_limiter.go

```diff
@@ -131,12 +133,21 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
 		close(waitExpired)
 	})
 	defer timer.Stop()
-	select {
-	case <-ctx.Done():
-		reservation.Cancel()
-		return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
-	case <-waitExpired:
-		return nil
+
+	for {
+		select {
+		case <-ctx.Done():
+			reservation.Cancel()
+			return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
+		case <-waitExpired:
+			return nil
+		case <-l.recycleCh:
+			if token > 1 {
+				break
```

Comment:
My intention here is to make RecycleToken only unblock a waiter if the waiter is asking for 1 token, since I felt like that behavior would be easier to guarantee and test, and it's rare that we ask for more than one rate limit token in temporal in general, so this behavior covers most use cases. All the rate limit waits in matching are only using one token, so this is good enough there.

Reply:
Looks like you read the other comment thread about this and agree as long as the behavior is commented, which it is! Resolving this comment for bookkeeping hope that's alright.

Reply:
yeah, my point was that we could make the receiving side logic correct, without worrying about the sending side yet. I admit it's a little silly since it's not tested. so either way is fine